File: plugins/wise/engine/src/adapters/gemini.ts
// `gemini` harness adapter (research-ts-engine.md P6, D15; plan M5.3). Best effort: the CLI is
// installed here (0.46.0) but the user's login is broken (see probeAuth), so the parser is built
// from the shipped bundle's source and the `--help` text rather than a live run.
//
// Flags confirmed on gemini 0.46.0 (`gemini --help` plus the bundled source):
//   -p/--prompt <text>        headless mode; "appended to input on stdin (if any)"
//   -o/--output-format        text | json | stream-json
//   -m/--model <name>
//   --approval-mode           default | auto_edit | yolo | plan  (-y/--yolo = yolo)
//   --include-directories     extra workspace dirs (repeatable or comma-separated)
//   -r/--resume <id>          "latest", a 1-based index, or a full session UUID (per project cwd)
//   --session-id <uuid>       start a fresh session under a chosen id
//   --skip-trust              required headless in a dir the user never trusted (else exit 55)
//   no effort / thinking flag, no schema flag, no max-turns flag, no system-prompt flag
// stream-json events (one JSON object per line): `init {session_id, model}`, `message {role,
// content, delta?}`, `tool_use {tool_name, tool_id, parameters}`, `tool_result {tool_id, status,
// output, error?}`, `error` (either `{severity, message}` for a non-fatal notice or
// `{error: {type, message, code?}, stats}` for a fatal one), `result {status: "success", stats}`
// with `stats {total_tokens, input_tokens, output_tokens, cached, input, duration_ms,
// tool_calls, models}`. No cost field anywhere. Fatal exit codes: 41 auth, 42 input, 53 turn
// limit, 54 tool execution, 55 untrusted workspace, 130 cancelled; API errors reuse the HTTP
// status (429). An auth failure before the runner starts prints plain text on stderr and exits 1.
// Stdin: when it is not a TTY the CLI reads it (up to 8 MiB, giving up after 500 ms of silence)
// and uses it as the prompt, with any `-p` text appended. A prompt over `PROMPT_ARGV_MAX` bytes
// therefore travels on stdin with no `-p`; smaller prompts go on `-p` and stdin is closed at once.
// Schema: no flag exists, so the schema is appended to the prompt as an instruction and the JSON
// object is extracted from the final assistant text (fenced block or bare object, with a warning).
// Auth files: `$GEMINI_CLI_HOME/.gemini/oauth_creds.json` (default `~/.gemini`) for the OAuth
// login, `google_accounts.json` beside it; `GOOGLE_API_KEY` beats `GEMINI_API_KEY`. There is no
// login subcommand or status command: login happens in the interactive `gemini` via `/auth`.

import { readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { effortFor } from "../resolve.ts";
import type {
  Adapter,
  AuthMode,
  Effort,
  RawEvent,
  RunMode,
  RunReq,
  RunRes,
  Usage,
} from "../types.ts";
import { cleanEnv, createLineSplitter, spawnClean } from "./spawn.ts";
import type { SpawnExit } from "./spawn.ts";

export const GEMINI_BIN = "gemini";
/** Either key authenticates; the CLI prefers `GOOGLE_API_KEY` when both are set. */
export const GEMINI_KEY_VARS = ["GOOGLE_API_KEY", "GEMINI_API_KEY"] as const;
export const GEMINI_CONFIG_VAR = "GEMINI_CLI_HOME";
/** Copied when set: Code Assist accounts on a paid tier need the project id. */
export const GEMINI_KEEP_VARS = [GEMINI_CONFIG_VAR, "GOOGLE_CLOUD_PROJECT"] as const;

/**
 * D19: `yolo` only under `full-access`. Headless gemini cannot answer a confirmation, so under
 * `default` any tool that would ask is refused with a FatalToolExecutionError (exit 54).
 */
export const APPROVAL_MAP: Record<RunMode, string> = {
  "approval-required": "default",
  auto: "auto_edit",
  "full-access": "yolo",
};

export const RATE_LIMIT_RE = /rate.?limit|429|too many requests|quota|RESOURCE_EXHAUSTED/i;
export const AUTH_RE =
  /error authenticating|IneligibleTier|FatalAuthenticationError|not logged in|unauthenticated|unauthorized|401|invalid api key|API key not valid|credentials|log ?in|authenticate/i;
export const TURN_LIMIT_RE = /FatalTurnLimitedError|turn limit|max.?turns|session turns/i;
const ERROR_TEXT_MAX = 500;

// ---- argv ---------------------------------------------------------------------------------------

/** Prompts above this many bytes go on stdin (no `-p`) instead of argv. */
export const PROMPT_ARGV_MAX = 100_000;

export function promptViaStdin(prompt: string): boolean {
  return Buffer.byteLength(prompt, "utf8") > PROMPT_ARGV_MAX;
}

const SCHEMA_INSTRUCTION =
  "Respond with only a JSON object matching this JSON schema. No prose, no code fence, nothing before or after the object.";

/**
 * The prompt gemini receives: the system text has no flag, so it leads; the schema has no flag,
 * so it trails as an instruction (schema-by-instruction).
 */
export function composePrompt(req: Pick<RunReq, "prompt" | "system" | "schema">): string {
  const parts: string[] = [];
  if (req.system !== undefined) parts.push(req.system);
  parts.push(req.prompt);
  if (req.schema !== undefined) parts.push(`${SCHEMA_INSTRUCTION}\n${JSON.stringify(req.schema)}`);
  return parts.join("\n\n");
}

export function buildArgv(req: RunReq): string[] {
  const argv: string[] = [];
  const prompt = composePrompt(req);
  if (!promptViaStdin(prompt)) argv.push("-p", prompt);
  argv.push("--output-format", "stream-json", "--skip-trust");
  argv.push("--approval-mode", APPROVAL_MAP[req.mode]);
  for (const dir of req.add_dirs ?? []) argv.push("--include-directories", dir);
  if (req.model && req.model !== "inherit") argv.push("-m", req.model);
  // No effort, max-turns or schema flag on this CLI: `effortMap` is undefined for every level.
  if (typeof req.resume === "string" && req.resume.length > 0) argv.push("--resume", req.resume);
  return argv;
}

export function childEnv(
  req: Pick<RunReq, "auth" | "env">,
  parent?: NodeJS.ProcessEnv,
): Record<string, string> {
  // An API key in the environment beats the cached OAuth login inside gemini, so both key
  // variables travel only under `api-key`.
  return cleanEnv({
    ...(parent ? { parent } : {}),
    keep: GEMINI_KEEP_VARS,
    secrets: req.auth === "api-key" ? GEMINI_KEY_VARS : [],
    ...(req.env ? { extra: req.env } : {}),
  });
}

export function effortMap(e: Effort): string | undefined {
  return effortFor("gemini", e);
}

// ---- schema-by-instruction extraction -------------------------------------------------------------

type Rec = Record<string, unknown>;
const isRec = (v: unknown): v is Rec => typeof v === "object" && v !== null && !Array.isArray(v);
const str = (v: unknown): string | undefined => (typeof v === "string" ? v : undefined);
const num = (v: unknown): number | undefined =>
  typeof v === "number" && Number.isFinite(v) ? v : undefined;

export type Extracted =
  | { ok: true; json: unknown; warning?: string }
  | { ok: false; error: string };

const FENCE_RE = /```(?:json|JSON)?\s*\n([\s\S]*?)\n\s*```/;

/** Index one past the `}` closing the object that opens at `start`, string-aware; -1 if none. */
function objectEnd(text: string, start: number): number {
  let depth = 0;
  let inString = false;
  for (let i = start; i < text.length; i++) {
    const ch = text[i];
    if (inString) {
      if (ch === "\\") i++;
      else if (ch === '"') inString = false;
    } else if (ch === '"') inString = true;
    else if (ch === "{") depth++;
    else if (ch === "}") {
      depth--;
      if (depth === 0) return i + 1;
    }
  }
  return -1;
}

const tryParse = (s: string): { ok: true; value: unknown } | { ok: false } => {
  try {
    return { ok: true, value: JSON.parse(s) };
  } catch {
    return { ok: false };
  }
};

/**
 * The model was asked for a bare JSON object; accept the whole text, a fenced block, or the first
 * balanced `{…}` inside prose, warning when anything but the whole text parsed.
 */
export function extractJson(text: string): Extracted {
  const whole = tryParse(text.trim());
  if (whole.ok) return { ok: true, json: whole.value };
  const fence = FENCE_RE.exec(text);
  if (fence?.[1] !== undefined) {
    const inner = tryParse(fence[1].trim());
    if (inner.ok) {
      return { ok: true, json: inner.value, warning: "JSON extracted from a fenced code block" };
    }
  }
  let from = text.indexOf("{");
  while (from !== -1) {
    const end = objectEnd(text, from);
    if (end === -1) break;
    const bare = tryParse(text.slice(from, end));
    if (bare.ok) {
      return { ok: true, json: bare.value, warning: "JSON extracted from surrounding prose" };
    }
    from = text.indexOf("{", from + 1);
  }
  const shown = text.trim();
  return {
    ok: false,
    error: `final assistant text is not JSON: ${shown.length > ERROR_TEXT_MAX ? shown.slice(0, ERROR_TEXT_MAX) : shown || "(empty)"}`,
  };
}

// ---- stream parser ------------------------------------------------------------------------------

export type StreamSnapshot = {
  session_id?: string;
  model?: string;
  /** Assistant text segments so far; a segment closes at each tool result. */
  segments: number;
  tool_uses: number;
  /** A `result` event arrived. */
  result: boolean;
  errors: string[];
  warnings: string[];
};

export type StreamParser = {
  feed(chunk: string): RawEvent[];
  finish(exit: SpawnExit): RunRes;
  snapshot(): StreamSnapshot;
};

export type ParserOpts = {
  pool: AuthMode;
  /** A schema was given: the final assistant text must yield a JSON object. */
  expectJson?: boolean;
  now?: () => string;
};

/** stream-json `result.stats`; the per-model breakdown is ignored (totals are enough). */
function usageOf(stats: Rec | undefined, pool: AuthMode): Usage {
  const s = stats ?? {};
  return {
    input: num(s.input_tokens) ?? 0,
    output: num(s.output_tokens) ?? 0,
    cache_read: num(s.cached) ?? 0,
    cache_write: 0,
    pool,
  };
}

function clip(text: string): string {
  const t = text.trim();
  return t.length > ERROR_TEXT_MAX ? t.slice(0, ERROR_TEXT_MAX) : t;
}

/** Message of a fatal `error` event (`error: {type, message, code}`), tagged with its type. */
function fatalText(ev: Rec): string | undefined {
  if (!isRec(ev.error)) return undefined;
  const type = str(ev.error.type);
  const message = str(ev.error.message) ?? JSON.stringify(ev.error);
  const code = ev.error.code;
  const codeText = code === undefined ? "" : ` (code ${String(code)})`;
  return type ? `${type}: ${message}${codeText}` : `${message}${codeText}`;
}

export function createStreamParser(opts: ParserOpts): StreamParser {
  const lines = createLineSplitter();
  const now = opts.now ?? (() => new Date().toISOString());
  const snap: StreamSnapshot = {
    segments: 0,
    tool_uses: 0,
    result: false,
    errors: [],
    warnings: [],
  };
  const segments: string[] = [];
  let current = "";
  let stats: Rec | undefined;
  let status: string | undefined;
  let stdout = "";

  const closeSegment = (): void => {
    if (current.trim().length > 0) {
      segments.push(current);
      snap.segments = segments.length;
    }
    current = "";
  };

  const ingest = (line: string): RawEvent => {
    const ev: RawEvent = { ts: now(), harness: "gemini", line };
    if (line.trim().length === 0) return ev;
    let parsed: unknown;
    try {
      parsed = JSON.parse(line);
    } catch {
      return ev;
    }
    ev.parsed = parsed;
    if (!isRec(parsed)) return ev;
    const type = parsed.type;
    if (type === "init") {
      const sid = str(parsed.session_id);
      if (sid !== undefined) snap.session_id = sid;
      const model = str(parsed.model);
      if (model !== undefined) snap.model = model;
    } else if (type === "message") {
      if (parsed.role === "assistant") current += str(parsed.content) ?? "";
    } else if (type === "tool_use") {
      snap.tool_uses += 1;
    } else if (type === "tool_result") {
      closeSegment();
    } else if (type === "error") {
      const fatal = fatalText(parsed);
      if (fatal !== undefined) {
        snap.errors.push(fatal);
        if (isRec(parsed.stats)) stats = parsed.stats;
      } else {
        const message = str(parsed.message) ?? "gemini error event";
        if (parsed.severity === "error") snap.errors.push(message);
        else snap.warnings.push(message);
      }
    } else if (type === "result") {
      snap.result = true;
      status = str(parsed.status);
      if (isRec(parsed.stats)) stats = parsed.stats;
    }
    return ev;
  };

  const classify = (exit: SpawnExit): Pick<RunRes, "exit" | "error"> => {
    const stderr = exit.stderr;
    if (exit.timedOut) return { exit: "timeout", error: clip(stderr) || "timed out" };
    if (snap.result && status === "success") return { exit: "ok" };
    const failure = snap.errors.at(-1);
    const haystack = `${snap.errors.join("\n")}\n${stderr}`;
    const detail =
      failure ??
      exit.error ??
      (stderr.trim()
        ? clip(stderr)
        : stdout.trim()
          ? `no result event (stdout: ${clip(stdout)})`
          : `no result event (exit code ${String(exit.code)}, signal ${String(exit.signal)})`);
    if (exit.code === 53 || TURN_LIMIT_RE.test(haystack)) {
      return { exit: "max_turns", error: clip(detail) };
    }
    if (exit.code === 41 || AUTH_RE.test(haystack)) return { exit: "auth", error: clip(detail) };
    if (RATE_LIMIT_RE.test(haystack)) return { exit: "rate_limited", error: clip(detail) };
    return { exit: "error", error: clip(detail) };
  };

  return {
    feed(chunk) {
      stdout += chunk;
      return lines.feed(chunk).map(ingest);
    },
    snapshot: () => ({ ...snap, errors: [...snap.errors], warnings: [...snap.warnings] }),
    finish(exit) {
      lines.finish().forEach(ingest);
      closeSegment();
      const verdict = classify(exit);
      const res: RunRes = {
        text: segments.at(-1) ?? "",
        usage: usageOf(stats, opts.pool),
        exit: verdict.exit,
      };
      if (snap.session_id !== undefined) res.cursor = snap.session_id;
      if (verdict.error !== undefined) res.error = verdict.error;
      const warnings = [...snap.warnings];
      if (opts.expectJson && verdict.exit === "ok") {
        const extracted = extractJson(res.text);
        if (extracted.ok) {
          res.json = extracted.json;
          if (extracted.warning !== undefined) warnings.push(extracted.warning);
        } else {
          res.exit = "error";
          res.error = extracted.error;
        }
      }
      if (warnings.length > 0) res.warnings = warnings;
      return res;
    },
  };
}

// ---- process lifecycle --------------------------------------------------------------------------

/** Same shape as `ClaudeRun` minus `nudge`: gemini reads stdin once, before the first turn. */
export type GeminiRun = {
  pid: number;
  done: Promise<RunRes>;
  kill(signal?: NodeJS.Signals): void;
  snapshot(): StreamSnapshot;
};

export type StartOpts = { bin?: string; parentEnv?: NodeJS.ProcessEnv; home?: string };

export function startGemini(
  req: RunReq,
  onEvent: (e: RawEvent) => void,
  opts: StartOpts = {},
): GeminiRun {
  const argv = buildArgv(req);
  const proc = spawnClean(opts.bin ?? GEMINI_BIN, argv, {
    cwd: req.cwd,
    env: childEnv(req, opts.parentEnv),
    timeoutMs: req.timeout_ms,
  });
  // The CLI reads a piped stdin as the prompt (giving up after 500 ms of silence): hand it the
  // whole prompt when argv would be too large, otherwise close it before the first byte.
  const prompt = composePrompt(req);
  if (promptViaStdin(prompt)) proc.stdin.end(prompt);
  else proc.stdin.end();
  const parser = createStreamParser({ pool: req.auth, expectJson: req.schema !== undefined });
  proc.stdout.on("data", (chunk: string) => {
    for (const ev of parser.feed(chunk)) onEvent(ev);
  });
  const done = proc.exited.then((exit) => {
    const res = parser.finish(exit);
    if (req.resume !== undefined && typeof req.resume !== "string") {
      res.warnings = [...(res.warnings ?? []), "ignored non-string resume cursor"];
    }
    return res;
  });
  return {
    pid: proc.pid,
    done,
    kill: (signal) => proc.kill(signal),
    snapshot: () => parser.snapshot(),
  };
}

// ---- adapter ------------------------------------------------------------------------------------

/** `$GEMINI_CLI_HOME/.gemini/oauth_creds.json`, default `~/.gemini/oauth_creds.json`. */
export function authFilePath(env: NodeJS.ProcessEnv = process.env, home?: string): string {
  const root = env[GEMINI_CONFIG_VAR] ?? home ?? env.HOME ?? homedir();
  return join(root, ".gemini", "oauth_creds.json");
}

/**
 * gemini 0.46.0 has no login-status command, so the subscription probe is a file check: an
 * `oauth_creds.json` holding an access or refresh token counts as logged in. An unexpired token
 * on an ineligible account (this machine: "IneligibleTierError: This client is no longer
 * supported for Gemini Code Assist for individuals") passes the probe and fails the run with an
 * `auth` exit instead.
 */
export async function probeAuth(
  auth: AuthMode,
  opts: StartOpts = {},
): Promise<{ ok: boolean; login_cmd?: string }> {
  const env = opts.parentEnv ?? process.env;
  if (auth === "api-key") {
    const ok = GEMINI_KEY_VARS.some((v) => Boolean(env[v]));
    return { ok, login_cmd: `export ${GEMINI_KEY_VARS[1]}=...` };
  }
  let ok = false;
  try {
    const parsed: unknown = JSON.parse(readFileSync(authFilePath(env, opts.home), "utf8"));
    ok = isRec(parsed) && (Boolean(str(parsed.access_token)) || Boolean(str(parsed.refresh_token)));
  } catch {
    ok = false;
  }
  return { ok, login_cmd: "gemini (interactive, then /auth)" };
}

export const geminiAdapter: Adapter = {
  id: "gemini",
  bin: GEMINI_BIN,
  probeAuth: (auth) => probeAuth(auth),
  run: (req, onEvent) => startGemini(req, onEvent).done,
  effortMap,
};

File: plugins/wise/engine/wise_engine/adapters/gemini.py
from __future__ import annotations

import json
import os
import re
import sys
import uuid
from collections.abc import Mapping
from pathlib import Path
from typing import Any

from ..adapter_types import AgentHandle, EventCallback, Json
from ..host_setup import _json_clean
from ..resolve import effort_for
from ..spawn import SpawnExit, SpawnOptions, clean_env, spawn_clean
from ._common import (
    Parser,
    ProviderAdapter,
    TemporaryFile,
    arm_task,
    clip,
    dumps,
    exit_detail,
    finish_process,
    loads,
    number,
    rec,
    string,
    temporary_file,
)

GEMINI_BIN = "gemini"
GEMINI_KEY_VARS = ("GOOGLE_API_KEY", "GEMINI_API_KEY")
GEMINI_CONFIG_VAR = "GEMINI_CLI_HOME"
GEMINI_SYSTEM_SETTINGS = "GEMINI_CLI_SYSTEM_SETTINGS_PATH"
GEMINI_SYSTEM_DEFAULTS = "GEMINI_CLI_SYSTEM_DEFAULTS_PATH"
GEMINI_KEEP_VARS = (
    GEMINI_CONFIG_VAR,
    "GOOGLE_CLOUD_PROJECT",
    GEMINI_SYSTEM_SETTINGS,
    GEMINI_SYSTEM_DEFAULTS,
)
APPROVAL_MAP = {"approval-required": "default", "auto": "auto_edit", "full-access": "yolo"}
RATE_LIMIT_RE = re.compile(r"rate.?limit|429|too many requests|quota|RESOURCE_EXHAUSTED", re.I)
AUTH_RE = re.compile(
    r"error authenticating|IneligibleTier|FatalAuthenticationError|not logged in|unauthenticated|unauthorized|401|invalid api key|API key not valid|credentials|log ?in|authenticate",
    re.I,
)
TURN_LIMIT_RE = re.compile(r"FatalTurnLimitedError|turn limit|max.?turns|session turns", re.I)
PROMPT_ARGV_MAX = 100_000
SCHEMA_INSTRUCTION = "Respond with only a JSON object matching this JSON schema. No prose, no code fence, nothing before or after the object."
FENCE_RE = re.compile(r"```(?:json|JSON)?\s*\n([\s\S]*?)\n\s*```")


def prompt_via_stdin(prompt: str) -> bool:
    return len(prompt.encode("utf-8")) > PROMPT_ARGV_MAX


def compose_prompt(req: Json) -> str:
    parts = [req["system"]] if "system" in req else []
    parts.append(req["prompt"])
    if "schema" in req:
        parts.append(SCHEMA_INSTRUCTION + "\n" + dumps(req["schema"]))
    return "\n\n".join(parts)


def build_argv(req: Json) -> list[str]:
    prompt = compose_prompt(req)
    argv = [] if prompt_via_stdin(prompt) else ["-p", prompt]
    argv += [
        "--output-format",
        "stream-json",
        "--skip-trust",
        "--approval-mode",
        APPROVAL_MAP[req["mode"]],
    ]
    for directory in req.get("add_dirs", []):
        argv += ["--include-directories", directory]
    if req.get("model") and req["model"] != "inherit":
        argv += ["-m", req["model"]]
    if isinstance(req.get("resume"), str) and req["resume"]:
        argv += ["--resume", req["resume"]]
    return argv


def child_env(req: Json, parent: Mapping[str, str | None] | None = None) -> dict[str, str]:
    return clean_env(
        parent=parent,
        keep=GEMINI_KEEP_VARS,
        secrets=GEMINI_KEY_VARS if req["auth"] == "api-key" else [],
        extra=req.get("env"),
    )


def prepare_mcp(req: Json, env: dict[str, str]) -> TemporaryFile | None:
    config = req.get("mcp_config", {})
    if not isinstance(config, dict) or not isinstance(config.get("mcpServers", {}), dict):
        raise ValueError("gemini: mcp_config must contain an MCP server mapping")
    servers = config.get("mcpServers", {})
    if not servers:
        return None
    for server in servers.values():
        if (
            not isinstance(server, dict)
            or not isinstance(server.get("command"), str)
            or not server["command"]
            or not isinstance(server.get("args", []), list)
            or not all(isinstance(arg, str) for arg in server.get("args", []))
            or not isinstance(server.get("env", {}), dict)
            or not all(
                isinstance(key, str) and isinstance(value, str)
                for key, value in server.get("env", {}).items()
            )
        ):
            raise ValueError(
                "gemini: child MCP requires a stdio command, string arguments and string environment"
            )
    default = (
        "/Library/Application Support/GeminiCli/settings.json"
        if sys.platform == "darwin"
        else "/etc/gemini-cli/settings.json"
    )
    original = Path(env.get(GEMINI_SYSTEM_SETTINGS, default)).expanduser().absolute()
    try:
        settings = json.loads(_json_clean(original.read_text()))
    except FileNotFoundError:
        settings = {}
    if not isinstance(settings, dict) or not isinstance(settings.get("mcpServers", {}), dict):
        raise ValueError("gemini: system settings must contain a valid MCP server mapping")
    policy = settings.get("mcp", {})
    if isinstance(policy, dict) and policy.get("allowed") is not None:
        raise ValueError("gemini: system MCP allowlist prevents injecting a per-step server")
    merged = dict(settings.get("mcpServers", {}))
    for server in servers.values():
        alias = "wise-step-" + uuid.uuid4().hex
        values = {}
        for index, (key, value) in enumerate(server.get("env", {}).items()):
            name = "WISE_MCP_" + uuid.uuid4().hex.upper() + "_" + str(index)
            env[name] = value
            values[key] = "${" + name + "}"
        merged[alias] = {**server, "env": values, "trust": True}
    overlay = temporary_file(
        "wise-gemini-mcp-", "settings.json", dumps({**settings, "mcpServers": merged})
    )
    try:
        Path(overlay.path).chmod(0o600)
    except BaseException:
        overlay.cleanup()
        raise
    env[GEMINI_SYSTEM_SETTINGS] = overlay.path
    # Keep defaults at their original location when the override file moves.
    env.setdefault(GEMINI_SYSTEM_DEFAULTS, str(original.with_name("system-defaults.json")))
    return overlay


def effort_map(effort: str) -> str | None:
    return effort_for("gemini", effort)


def extract_json(text: str) -> Json:
    try:
        return dict(ok=True, json=loads(text.strip()))
    except ValueError:
        pass
    fence = FENCE_RE.search(text)
    if fence:
        try:
            return dict(
                ok=True,
                json=loads(fence[1].strip()),
                warning="JSON extracted from a fenced code block",
            )
        except ValueError:
            pass
    start = text.find("{")
    while start != -1:
        depth = 0
        in_string = False
        index = start
        end = -1
        while index < len(text):
            char = text[index]
            if in_string:
                if char == "\\":
                    index += 1
                elif char == '"':
                    in_string = False
            elif char == '"':
                in_string = True
            elif char == "{":
                depth += 1
            elif char == "}":
                depth -= 1
                if depth == 0:
                    end = index + 1
                    break
            index += 1
        if end == -1:
            break
        try:
            return dict(
                ok=True,
                json=loads(text[start:end]),
                warning="JSON extracted from surrounding prose",
            )
        except ValueError:
            start = text.find("{", start + 1)
    return dict(ok=False, error=f"final assistant text is not JSON: {clip(text) or '(empty)'}")


class StreamParser(Parser):
    harness = "gemini"

    def __init__(self, *, expect_json: bool = False, **opts: Any) -> None:
        super().__init__(**opts)
        self.expect_json = expect_json
        self.snap = dict(segments=0, tool_uses=0, result=False, errors=[], warnings=[])
        self.segments: list[str] = []
        self.current = ""
        self.stats: Json = {}
        self.status: str | None = None
        self.stdout = ""

    def close_segment(self) -> None:
        if self.current.strip():
            self.segments.append(self.current)
            self.snap["segments"] = len(self.segments)
        self.current = ""

    def feed(self, chunk: str | bytes) -> list[Json]:
        text = self.decode_chunk(chunk)
        self.stdout += text
        return super().feed(text)

    def accept(self, parsed: Json) -> None:
        kind = parsed.get("type")
        if kind == "init":
            for name in ("session_id", "model"):
                if isinstance(parsed.get(name), str):
                    self.snap[name] = parsed[name]
        elif kind == "message" and parsed.get("role") == "assistant":
            self.current += string(parsed.get("content")) or ""
        elif kind == "tool_use":
            self.snap["tool_uses"] += 1
        elif kind == "tool_result":
            self.close_segment()
        elif kind == "error":
            error = parsed.get("error")
            if isinstance(error, dict):
                message = string(error.get("message"))
                if message is None:
                    message = dumps(error)
                code = f" (code {error['code']})" if "code" in error else ""
                self.snap["errors"].append(
                    f"{error['type']}: {message}{code}"
                    if isinstance(error.get("type"), str) and error["type"]
                    else message + code
                )
                if isinstance(parsed.get("stats"), dict):
                    self.stats = parsed["stats"]
            else:
                message = string(parsed.get("message"))
                self.snap["errors" if parsed.get("severity") == "error" else "warnings"].append(
                    message if message is not None else "gemini error event"
                )
        elif kind == "result":
            self.snap["result"] = True
            self.status = string(parsed.get("status"))
            if isinstance(parsed.get("stats"), dict):
                self.stats = parsed["stats"]

    def classify(self, exit: SpawnExit) -> Json:
        if exit.timed_out:
            return dict(exit="timeout", error=clip(exit.stderr) or "timed out")
        if self.snap["result"] and self.status == "success":
            return dict(exit="ok")
        errors = self.snap["errors"]
        haystack = "\n".join(errors) + "\n" + exit.stderr
        detail = (
            errors[-1]
            if errors
            else exit.error
            if exit.error is not None
            else clip(exit.stderr)
            if exit.stderr.strip()
            else f"no result event (stdout: {clip(self.stdout)})"
            if self.stdout.strip()
            else exit_detail(exit, "no result event")
        )
        kind = (
            "max_turns"
            if exit.code == 53 or TURN_LIMIT_RE.search(haystack)
            else "auth"
            if exit.code == 41 or AUTH_RE.search(haystack)
            else "rate_limited"
            if RATE_LIMIT_RE.search(haystack)
            else "error"
        )
        return dict(exit=kind, error=clip(detail))

    def finish(self, exit: SpawnExit) -> Json:
        self.flush()
        self.close_segment()
        usage = dict(
            input=number(self.stats.get("input_tokens")) or 0,
            output=number(self.stats.get("output_tokens")) or 0,
            cache_read=number(self.stats.get("cached")) or 0,
            cache_write=0,
            pool=self.pool,
        )
        res = dict(
            text=self.segments[-1] if self.segments else "", usage=usage, **self.classify(exit)
        )
        if "session_id" in self.snap:
            res["cursor"] = self.snap["session_id"]
        warnings = list(self.snap["warnings"])
        if self.expect_json and res["exit"] == "ok":
            extracted = extract_json(res["text"])
            if extracted["ok"]:
                res["json"] = extracted["json"]
                if "warning" in extracted:
                    warnings.append(extracted["warning"])
            else:
                res.update(exit="error", error=extracted["error"])
        if warnings:
            res["warnings"] = warnings
        return res


def create_stream_parser(**opts: Any) -> StreamParser:
    return StreamParser(**opts)


async def start_gemini(
    req: Json,
    on_event: EventCallback,
    *,
    bin: str = GEMINI_BIN,
    parent_env: Mapping[str, str | None] | None = None,
) -> AgentHandle:
    env = child_env(req, parent_env)
    overlay = prepare_mcp(req, env)
    try:
        proc = await spawn_clean(
            bin,
            build_argv(req),
            SpawnOptions(cwd=req["cwd"], env=env, timeout_ms=req["timeout_ms"]),
        )
    except BaseException:
        if overlay:
            overlay.cleanup()
        raise
    prompt = compose_prompt(req)
    proc.stdin.end(prompt if prompt_via_stdin(prompt) else "")
    parser = StreamParser(pool=req["auth"], expect_json="schema" in req)

    async def finish() -> Json:
        res = await finish_process(proc, parser, on_event, overlay.cleanup if overlay else None)
        if "resume" in req and not isinstance(req["resume"], str):
            res["warnings"] = [*res.get("warnings", []), "ignored non-string resume cursor"]
        return res

    return AgentHandle(
        done=await arm_task(finish(), proc, overlay.cleanup if overlay else None),
        pid=proc.pid,
        kill=proc.kill,
        snapshot=parser.snapshot,
    )


def auth_file_path(env: Mapping[str, str | None] | None = None, home: str | None = None) -> str:
    env = os.environ if env is None else env
    root = env.get(GEMINI_CONFIG_VAR)
    if root is None:
        root = home if home is not None else env.get("HOME")
    return str(Path(root if root is not None else Path.home()) / ".gemini" / "oauth_creds.json")


async def probe_auth(
    auth: str, *, parent_env: Mapping[str, str | None] | None = None, home: str | None = None
) -> Json:
    env = os.environ if parent_env is None else parent_env
    if auth == "api-key":
        return dict(
            ok=any(env.get(name) for name in GEMINI_KEY_VARS), login_cmd="export GEMINI_API_KEY=..."
        )
    try:
        parsed = rec(loads(Path(auth_file_path(env, home)).read_text(encoding="utf-8")))
        ok = bool(string(parsed.get("access_token")) or string(parsed.get("refresh_token")))
    except (OSError, ValueError):
        ok = False
    return dict(ok=ok, login_cmd="gemini (interactive, then /auth)")


gemini_adapter = ProviderAdapter("gemini", GEMINI_BIN, start_gemini, probe_auth, effort_map)
